mesonbuild/backend/ninjabackend.py

```diff
@@ -1103,6 +1103,22 @@ def generate_vala_compile(self, target, outfile):
                 # Install GIR to default location if requested by user
                 if len(target.install_dir) > 3 and target.install_dir[3] is True:
                     target.install_dir[3] = os.path.join(self.environment.get_datadir(), 'gir-1.0')
+            # Generate header for internal symbols if requested by user
+            if isinstance(target.vala_internal_header, str):
+                internalheadername = os.path.join(self.get_target_dir(target), target.vala_internal_header)
+                args += ['--internal-header', internalheadername]
+                valac_outputs.append(internalheadername)
+                target.outputs.append(target.vala_internal_header)
+                if len(target.install_dir) > 4 and target.install_dir[4] is True:
+                    target.install_dir[4] = self.environment.get_includedir()
```

Comment:
Like any other C project, you have public API and protected API. The first is for your use without internals. The second is when you subclass a GObject and you want/need to access some internal structures not public but available in subclasses, this is why valac adds a --internal-header, making easy to C subclasses to access this protected structs/variables.

In Vala, when you declare a variable or class as *internal* they are not public and now you can get access it using a VAPI for internal definitions, like Gee does, so you can extend or use them using subclasses written in Vala.

Now is important to extend ninja backend to support generation of internal header and VAPI definitions, because they are produced  at first run valac call, installing them is important too if you want any other C or Vala coder to use them, so install script may is too much and is simpler to install them in $(include) and $(vapi) directories as public API is located.